k2/csrc/fsa_algo.cc

```diff
@@ -275,6 +275,111 @@ bool Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map /*=nullptr*/) {
   return is_acyclic;
 }
 
+void RmEpsilonsPrunedMax(const WfsaWithFbWeights &a, float beam, Fsa *b,
+                         std::vector<std::vector<int32_t>> *arc_derivs) {
+  CHECK_EQ(a.weight_type, kMaxWeight);
+  CHECK_GT(beam, 0);
+  CHECK_NOTNULL(b);
+  CHECK_NOTNULL(arc_derivs);
+  b->arc_indexes.clear();
+  b->arcs.clear();
+  arc_derivs->clear();
+
+  const auto &fsa = a.fsa;
+  if (IsEmpty(fsa)) return;
+  int32_t num_states_a = fsa.NumStates();
+  int32_t final_state = fsa.FinalState();
+  const auto &arcs_a = fsa.arcs;
+  const float *arc_weights_a = a.arc_weights;
+
+  // identify all states that should be kept
+  std::vector<char> non_eps_in(num_states_a, 0);
+  non_eps_in[0] = 1;
+  using WeightsPair = std::pair<double, std::vector<int32_t>>;
+  // (label, dest_state) -> `sum` of weights along all paths from current state
+  // to `dest_state` with label == `label`(or plus numbers of epsilon),
+  // `vector<int32_t>` in `WeightsPair` records all arc-indexes in `a` that
+  // contributes the current arc in `b`.
+  using ArcMap =
+      std::unordered_map<std::pair<int32_t, int32_t>, WeightsPair, PairHash>;
+  std::vector<ArcMap> arcs_b(num_states_a);
+  for (int32_t i = static_cast<int32_t>(arcs_a.size()) - 1; i >= 0; --i) {
+    const auto &arc = arcs_a[i];
+    const auto src_state = arc.src_state;
+    const auto dest_state = arc.dest_state;
+    const auto label = arc.label;
+    DCHECK_GE(dest_state, src_state);
```

Comment:
You are relying on topological sorting.. comment so near the top of this function and point the reader to this line where you check it.

Reply:
OK, will add in `.cc` file (actually you have commented in the header file to require TopSorted)
https://github.com/danpovey/k2/blob/master/k2/csrc/fsa_algo.h#L84-L87